group_vars/all/main.yml

```diff
@@ -35,16 +35,17 @@ backend_db_name: 'bluebutton_backend_sandbox'
 backend_fhir_port_internal: 8443
 backend_fhir_port_external: 443
 
-# 8x vCPUs, 32 GB RAM
+# 8x vCPUs, 32 GB RAM, 1065 max connections on Postgres 9.6 (per `select * from pg_settings where name='max_connections';`)
 rds_backend_db_instance_type: 'db.m4.2xlarge'
 
 # 16x vCPUs, 64 GB RAM.
-ec2_backend_fhir_instance_type: 'm4.4xlarge'
 ec2_backend_etl_instance_type: 'm4.4xlarge'
+ec2_backend_fhir_instance_type: 'm4.4xlarge'
+data_pipeline_jvm_args: '-Xmx58g'
+data_pipeline_loader_threads: 100  # The service will use x2 this number of DB connections.
 data_server_appserver_jvmargs: '-Xmx58g -XX:MaxMetaspaceSize=4g'
 data_server_db_max_connections: '100'
-data_pipeline_jvm_args: '-Xmx58g'
-data_pipeline_loader_threads: '390'
+data_server_asg_max_size: 7  # Watch out that this (plus extra instances for next deploy) doesn't exceed the max allowed DB connections!
```

Comment:
How did you come up with the number 7(looks like it was 10 before...)?

Reply:
Very good question!

Up above, I added a comment noting that the DB supports a max of 1065 connections. This playbook sets up systems that consume those connections, as follows:
* 200: (1 Always-On Data Pipeline service) x (`data_pipeline_loader_threads` threads per service) x (2 connections per thread)
* 700: (`data_server_asg_max_size` auto-scaled Data Server instances) x (`data_server_db_max_connections` connections per instance)
* 100: (1 additional temporary Data Server instance during deploys) x (`data_server_db_max_connections` connections per instance)

That adds up to a total of 1000 DB connections, which is close to the limit, but still leaves us a bit of room for normal admin/ops activity.